### PR TITLE
Assignment2 - Anna Mikhailenko and Sergei Ashikhmin

### PR DESCRIPTION
**Pull request by Anna-Mclenko**
Team: @Anna-Mclenko and @SergeiAshikhmin

Note: be cautious with unit tests. There is a file path string that is not relative, it was configured to match my Windows machine. Update it if you want to run my tests on your machine.

### CODE REVIEW

**Comment by Joshua-Lester3** [comment]
I will review this

**Review by Joshua-Lester3 — COMMENTED** [review]
There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ❌
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ❌
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔
All of the above should be unit tested.

File: Logger/LogFactory.cs [hunk 1]
```diff
@@ -1,10 +1,23 @@
...
+namespace Logger;
 
 public class LogFactory
 {
-    public BaseLogger CreateLogger(string className)
+    private string? _PathName;
+    public BaseLogger? CreateLogger(string className)
     {
+        if(className == nameof(FileLogger)){
+            FileLogger FileLogger = new(_PathName!) {ClassName = className};
```
[the remark this answers is not in the document]
Comment (Anna-Mclenko) [reply Logger/LogFactory.cs]: Oh thanks for catching that!

File: Logger.Tests/LogFactoryTests.cs [hunk 2]
```diff
@@ -5,5 +5,36 @@ namespace Logger.Tests;
 [TestClass]
 public class LogFactoryTests
```
[the remark this answers is not in the document]
Comment (Anna-Mclenko) [reply Logger.Tests/LogFactoryTests.cs:6]: I don't think that we need to test that. I think that It will be redundant because the className will always be the same name as the created object that is returned, if the class name provided exists. What we do need to test though is whether the createLogger method returns null when the classname provided does not exist.

File: Logger/FileLogger.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+{
+    private readonly string _PathName;
+    public FileLogger(string filePath)
+    {
+        _PathName = filePath;
+    }
+    public override void Log(LogLevel logLevel, string message)
+    {
+        DateTime DateTime = DateTime.Now;
+        string ClassName = this.ClassName ?? "Null name???";
```
[the remark this answers is not in the document]
Comment (Anna-Mclenko) [reply Logger/FileLogger.cs]: thanks for pointing this out!

File: Logger/LogFactory.cs [hunk 4]
```diff
@@ -1,10 +1,23 @@
...
+
+namespace Logger;
 
 public class LogFactory
 {
-    public BaseLogger CreateLogger(string className)
+    private string? _PathName;
+    public BaseLogger? CreateLogger(string className)
     {
+        if(className == nameof(FileLogger)){
```
[the remark this answers is not in the document]
Comment (Anna-Mclenko) [reply Logger/LogFactory.cs]: Oh! That makes more senese! Thank you! we fixed that!

File: Logger/LogFactory.cs [hunk 5]
```diff
@@ -1,10 +1,23 @@
...
+            FileLogger FileLogger = new(_PathName!) {ClassName = className};
+            return FileLogger;
+        }
 
         return null;
     }
+    public string ConfigureFileLogger(string pathName)
+        {
+            _PathName = pathName;
```
[the remark this answers is not in the document]
Comment (Anna-Mclenko) [reply Logger/LogFactory.cs]: Oh! thanks you got a point! we got that fixed

File: Logger/BaseLoggerMixins.cs [hunk 6]
```diff
@@ -1,6 +1,89 @@
...
 public static class BaseLoggerMixins
 {
+    public static void Error(this BaseLogger? logger, string message, params object[] args)
+    {
+        if(logger == null)
+        {
+            throw new ArgumentNullException(message);
+        }
+
+        if(args == null)
```
[the remark this answers is not in the document]
Comment (Anna-Mclenko) [reply Logger/BaseLoggerMixins.cs:15]: Its a feature. It appends all of the values from the args array to the message. since params argument can have anywhere from 0 to n values, we are simply just checking whether we can simply append a message if an array has not been provided or if we need to to format the message when the array is provided

**Comment by StrangeRanger** [comment]
@Joshua-Lester3 Make sure that all requirements that have been marked with an '❌', has it's appropriate code commented on.

Example:

> Use the nameof() operator when identifying the class name to the logger ❌

If you thought that the above requirement was not complete, make sure to mention in the code where `nameof()` should have been used.
